Add services for managing Time-of-Use (TOU) schedule for Growatt integration

johanzander opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

  Adds two new services to the Growatt Server integration for managing Time-of-Use
  (TOU) battery scheduling on MIN inverters:

  - `growatt_server.update_time_segment`: Configure individual time segments (1-9)
  with battery operation mode, time range, and enable/disable state
  - `growatt_server.read_time_segments`: Read current configuration of all 9 time
  segments

  These services enable users to automate their battery charging/discharging schedules
   based on electricity rates, solar generation, or other factors.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [X] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/41475

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.

  AI tools are welcome, but contributors are responsible for *fully*
  understanding the code before submitting a PR.
-->

- [X] I understand the code I am submitting and can explain how it works.
- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] I have followed the [perfect PR recommendations][perfect-pr]
- [X] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [X] Tests have been added to verify that the new code works.
- [X] Any generated code has been carefully reviewed for correctness and compliance with project standards.

If user exposed functionality or configuration variables are added/changed:

- [X] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

johanzander committed:
Add support for MIN time segment management in Growatt integration

johanzander committed:
Fix service tests after rebase

- Update fixture names from mock_growatt_api to mock_growatt_v1_api/mock_growatt_classic_api
- Fix min_write_time_segment mock to accept variable arguments
- Use correct Classic API fixture for non-MIN device test

johanzander committed:
Generalize service names for time segment management

Rename services from update_min_time_segment/read_min_time_segments
to update_time_segment/read_time_segments to support future expansion
to other inverter types beyond MIN devices.

johanzander committed:
Improve service code quality and test coverage

- Fix bare Exception catches in service handlers, replace with specific API exceptions
- Remove debug logging leftover from development
- Add comprehensive error handling tests for services
- Improve test coverage for service error paths
- All 42 tests passing with 85% overall coverage

johanzander committed:
Refactor test structure for improved readability and maintainability

johanzander committed:
Simplify return statement in _async_register_services function

joostlek commented on homeassistant/components/growatt_server/__init__.py:
```diff
@@ -185,9 +194,193 @@ async def async_setup_entry(
     await hass.config_entries.async_forward_entry_setups(config_entry, PLATFORMS)
 
+    # Register services for MIN/TLX devices
+    await _async_register_services(hass, config_entry, device_coordinators)
```
We should register this in `async_setup` instead so the availability of the services is not dependent on a working config entry

joostlek commented on homeassistant/components/growatt_server/__init__.py:
```diff
@@ -185,9 +194,193 @@ async def async_setup_entry(
     return True
 
 
+async def _async_register_services(
```
Consider moving this to a services.py

johanzander committed:
adress review comments

johanzander committed:
Restore number platform content in strings.json and conftest.py

johanzander committed:
Fix stale data issue: refresh coordinator after updating time segment

After writing a time segment value, the coordinator data was not refreshed,
causing read_time_segments to return stale/old values. Now triggers an
async_refresh() after successful write to ensure subsequent reads return
the updated values.

johanzander committed:
Revert "Fix stale data issue: refresh coordinator after updating time segment"

This reverts commit 830158d656374fd3e4a60c3f3088601887e5faf2.

johanzander committed:
Update coordinator cache after writing time segment to avoid stale data

After writing a time segment value, update the coordinator's cached data
directly without making an additional API call. This ensures that
subsequent read_time_segments calls return the updated values immediately
while avoiding rate limit issues.

Uses async_set_updated_data() to update the cache and notify entities of
the change without triggering a coordinator refresh.

johanzander committed:
Fix unique_id in conftest.py back to plant_123

Restore unique_id to 'plant_123' (was accidentally changed to '12345'
during merge conflict resolution). This matches the upstream/dev version
and the plant_id in the config data.

johanzander committed:
Merge branch 'dev' into growatt-services

johanzander committed:
Merge branch 'dev' into growatt-services

joostlek commented on homeassistant/components/growatt_server/services.py:
```diff
@@ -0,0 +1,181 @@
+
+        for entry in hass.config_entries.async_entries(DOMAIN):
+            if not entry.runtime_data:
+                continue
```
instead, check if the entry state is LOADED

joostlek commented on homeassistant/components/growatt_server/services.py:
```diff
@@ -0,0 +1,181 @@
+    # but will check for available coordinators when called
+    if hass.services.has_service(DOMAIN, "update_time_segment"):
+        # Services already registered
+        return
```
it won't ever re-register as async_setup is only called once in the HA lifetime

johanzander replied:
got it.

joostlek commented on homeassistant/components/growatt_server/strings.json:
```diff
@@ -523,5 +523,56 @@
+      "name": "Read time segments"
+    },
+    "update_time_segment": {
+      "description": "Update a time-of-use (TOU) segment for supported inverters.",
```
where does TOU come from? As in, is it a common term?

johanzander replied:
actually, I think I will change it to more generic time segment in the description, as these are used not only for TOU tariffs even though that was probably Growatts intention.

joostlek reviewed (changes requested):
Gave it a quick review, not in depth so might've missed some things

johanzander committed:
Adress review comments

johanzander committed:
Merge branch 'growatt-services' of https://github.com/johanzander/core into growatt-services

johanzander committed:
Merge branch 'dev' into growatt-services

johanzander commented on homeassistant/components/growatt_server/strings.json:
```diff
@@ -523,5 +523,56 @@
+        },
+        "device_id": {
+          "description": "Device ID (only required when multiple devices exist).",
+          "name": "Device ID"
```
done

johanzander commented on homeassistant/components/growatt_server/strings.json:
```diff
@@ -523,5 +523,56 @@
+      "description": "Read all time-of-use (TOU) segments from a supported inverter.",
+      "fields": {
+        "device_id": {
+          "description": "Device ID (only required when multiple devices exist).",
```
ok, done

johanzander committed:
Update description for device_id field in time segment service

johanzander committed:
Merge branch 'growatt-services' of https://github.com/johanzander/core into growatt-services

johanzander committed:
Fixes bug related to key names for TOU settings

johanzander committed:
Merge branch 'dev' into growatt-services

joostlek reviewed (commented):
Okay I need to look at this when I am not about to head to bed but I think this is a simple thing you can add to make the tests better :)

johanzander committed:
Add snapshot tests for reading time segments and update mock data

johanzander committed:
Merge branch 'growatt-services' of https://github.com/johanzander/core into growatt-services

johanzander committed:
Merge branch 'dev' into growatt-services

joostlek commented:
I'm on mobile now (I'm on holiday and I was reviewing this yesterday (or at least trying to) because I was bored). But I would need to check what the current parameters were.

In a way I would say that making the service generic enough to be used by both devices makes sense (as then blueprints and automations can be shared with a broader audience). But if the input parameters are drastically different then that probably doesn't make much sense.

So that would be my first preference, other than that, there are quite a number of growatt models, do we have any vision on if other inverters get the same? Or if they don't have it, do they carry the same type of parameters?

If the set of parameters in this PR is only specific for time segments and the one you are talking now is specific for charging ours, then it would make sense to have specific names for them, if we expect every model to have a wildly different set of parameters, having the model name in there would probably make sense (but now that I type this it probably isn't ideal either as you'd have "update segments (min)" "update segments (mix)" and whatnot, so when automating it can be harder to recognise what service to use when.

In any case, thanks for sharing your concern! I'll try to get back to this after I recover from my massive jet lag that I will have next weekend. Also, please don't forget to send me a message on discord so we can add you to the appropriate channels as codeowner :)

johanzander commented:
@joostlek - After checking all the different Growatt inverter APIs, I've found that there are 4 different ways they handle time-of-use/charge/discharge controls. Each type works completely differently.

## The 4 Types

### MIN Inverters - Time Segments

- 9 programmable time slots
- 3 modes: Load First, Battery First, Grid First
- No charge/discharge power settings (handled by separate number entities)

### NOAH Inverters - Basic Time Segments  

- 9 programmable time slots (like MIN)
- Only 2 modes: Load First, Battery First (no Grid First)
- Has output power limit setting per time segment

### MIX(SPH)/SPA Inverters - Charge/Discharge Periods

- Separate charge and discharge controls
- 3 time periods for each
- Power percentage controls, SOC limits, global on/off
- MIX and SPA work exactly the same

### STORAGE Inverters - Basic Periods

- Separate charge and discharge controls (like MIX)
- Only 1 time period for each
- Just simple on/off, no advanced settings

## Propsed Service Names

**Main models get clean names:**

MIN

- `growatt_server.update_time_segment`

MIX/SPA

- `growatt_server.update_charge_periods`
- `growatt_server.update_discharge_periods`

**Niche models get "basic_" prefix:**

NOAH

- `growatt_server.update_basic_time_segment`

STORAGE

- `growatt_server.update_basic_charge_period`
- `growatt_server.update_basic_discharge_period`

## Why This Makes Sense

- To the best of my knowledge, MIN and MIX inverters seem to be the newer more modern inverters with more advanced controls.
- NOAH and STORAGE are specialty products, so the "basic_" prefix makes it clear they're different
- Possible to create simple Blueprints based on these 4 different control types
- If a new inverter type is release, it can reuse any of the existing services (if parameters are same)

So to summarize, the service in this PR is unique for the min inverters, but I still think the naming makes sense.

johanzander commented:
yes, that it basically what I propose too. but not to prefix the inverter API type in the methods name, but name them for what they do (see above) - which is also what they are named in the PR. 

Another strong argument for this is because there is no logic on which inverter types use which API type.  I am about to update the documentation with which inverters use the MIN API, and it is a mix of MIC, MIN, MOD and MID: https://github.com/home-assistant/home-assistant.io/pull/41713.

If we would name the method min_update_time_segment, that would be very confusing for users of MIC, MOD and MID inverters...

johanzander committed:
Merge branch 'dev' into growatt-services

johanzander committed:
Merge branch 'dev' into growatt-services

johanzander committed:
Merge branch 'dev' into growatt-services

joostlek committed:
Update homeassistant/components/growatt_server/const.py
